bin/exec.ml

```diff
@@ -32,6 +32,75 @@ let man =
 
 let info = Cmd.info "exec" ~doc ~man
 
+module Command_to_exec = struct
+  (* A command to execute, which knows how to (re)build the program and then
+     run it with some arguments in an enivorment *)
+
+  type t =
+    { get_path_and_build_if_necessary :
+        unit -> (Path.t, [ `Already_reported ]) result Fiber.t
+    ; args : string list
+    ; env : Env.t
+    }
+
+  (* Helper function to spawn a new process running a command in an
+     environment, returning the new process' pid *)
+  let spawn_process path ~args ~env =
+    let path = Path.to_string path in
+    let env = Env.to_unix env |> Spawn.Env.of_list in
+    let argv = path :: args in
+    let pid = Spawn.spawn ~prog:path ~env ~argv () in
+    Pid.of_int pid
+
+  (* Run the command, first (re)building the program which the command is
+     invoking *)
+  let build_and_run_in_child_process
+      { get_path_and_build_if_necessary; args; env } =
+    get_path_and_build_if_necessary ()
+    |> Fiber.map ~f:(Result.map ~f:(spawn_process ~args ~env))
+end
+
+module Watch = struct
+  (* When running `dune exec` in watch mode, this will keep track of the pid of
+     the process created to run the program in the previous iteration so that
+     it can be killed (for long running programs, e.g. servers) and restarted
+     when its source is changed. *)
+
+  type state = { currently_running_pid : Pid.t option ref }
+
+  let init_state () = { currently_running_pid = ref None }
+
+  (* Helper function to kill a process and block the calling thread until the
+     killed process stops. The pid of the killed process will be reaped. *)
+  let kill_and_reap_process pid =
+    let pid_int = Pid.to_int pid in
+    Unix.kill pid_int Sys.sigkill;
+    let _stopped_pid, _process_status = Unix.waitpid [] pid_int in
```

Comment:
Why not use the Scheduler primitives for launching and waiting for processes?

Reply:
Does `Scheduler` provide a way of starting a process? The only relevant function I see is:
```ocaml
val wait_for_process :
     ?timeout:float
  -> ?is_process_group_leader:bool
  -> Pid.t
  -> Proc.Process_info.t Fiber.t
```
for waiting on a process. My goal is to start a process which runs concurrently to dune, and when a file is changed, to kill the child process if it's still running and then wait on it so its pid can be reaped. I don't see a way of doing this with the `Scheduler` module.

Reply:
There's `Process` for launching the process. But actually, you don't need that and you're free to use your existing function. You can definitely use `Scheduler.wait_for_process` though.

The other operations you've mentioned can be done outside of Scheduler.